### ISSUE TITLE
oasis.0.4.5 fails to install

### ISSUE THREAD

**Issue by amirmc** [ocaml/opam-repository#5172]
Not sure if this is just me. @gildor478 should I report this elsewhere?

```
=-=- Processing actions -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=  🐫
[ERROR] The compilation of oasis failed at "ocaml setup.ml -install".
[WARNING] Directory /Users/amir/.opam/mirage/lib/oasis is not empty, not removing

#=== ERROR while installing oasis.0.4.5 =======================================#
# opam-version 1.2.2
# os           darwin
# command      ocaml setup.ml -install
# path         /Users/amir/.opam/mirage/build/oasis.0.4.5
# compiler     system (4.02.3)
# exit-code    1
# env-file     /Users/amir/.opam/mirage/build/oasis.0.4.5/oasis-41192-c61e7b.env
# stdout-file  /Users/amir/.opam/mirage/build/oasis.0.4.5/oasis-41192-c61e7b.out
# stderr-file  /Users/amir/.opam/mirage/build/oasis.0.4.5/oasis-41192-c61e7b.err
### stderr ###
# File "src/oasis/OASISString.ml", line 148, characters 8-26:
# Warning 3: deprecated: String.set
# Use Bytes.set instead.
# File "src/oasis/OASISFindlib.ml", line 262, characters 6-24:
# Warning 3: deprecated: Lazy.lazy_from_fun
# Use Lazy.from_fun instead.
# ocamlfind: Package userconf is already installed
#  - (file /Users/amir/.opam/mirage/lib/userconf/META already exists)
# E: Failure("Command ''/Users/amir/.opam/mirage/bin/ocamlfind' install userconf src/ext/userconf/src/META _build/src/ext/userconf/src/UserConf.cmx _build/src/ext/userconf/src/UserConf.cmi _build/src/ext/userconf/src/userconf.cmxs _build/src/ext/userconf/src/userconf.a _build/src/ext/userconf/src/userconf.cmxa _build/src/ext/userconf/src/userconf.cma src/ext/userconf/src/UserConf.ml' terminated with error code 2")
```

```
# OPAM config report
# opam-version    1.2.2
# self-upgrade    no
# os              darwin
# external-solver aspcud $in $out $criteria
# criteria        -count(removed),-notuptodate(request),-sum(request,version-lag),-count(down),-notuptodate(changed),-count(changed),-notuptodate(solution),-sum(solution,version-lag)
# jobs            4
# repositories    1* (http)
# pinned          0
# current-switch  mirage*
# last-update     2015-11-24 12:16
```

**Comment by mor1** [comment]
Odd-- just tried this myself and no issues:

```
: mort@greyjay:papers#master *= $; opam remove oasis
[NOTE] oasis is not installed.

Nothing to do.
: mort@greyjay:papers#master *= $; opam install oasis
The following actions will be performed:
  ∗  install ocamlmod            0.0.8        [required by oasis]
  ∗  install ocaml-data-notation 0.0.11       [required by oasis]
  ∗  install ocamlify            0.0.1        [required by oasis]
  ∗  install oasis               0.4.5
===== ∗  4 =====
Do you want to continue ? [Y/n] y

=-=- Gathering sources =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=  🐫
[default] https://opam.ocaml.org/archives/ocamlify.0.0.1+opam.tar.gz downloaded
[default] https://opam.ocaml.org/archives/ocaml-data-notation.0.0.11+opam.tar.gz downloaded
[default] https://opam.ocaml.org/archives/oasis.0.4.5+opam.tar.gz downloaded
[default] https://opam.ocaml.org/archives/ocamlmod.0.0.8+opam.tar.gz downloaded

=-=- Processing actions -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=  🐫
∗  installed ocamlify.0.0.1
∗  installed ocamlmod.0.0.8
∗  installed ocaml-data-notation.0.0.11
∗  installed oasis.0.4.5
Done.
: mort@greyjay:papers#master *= $; opam update

=-=- Updating package repositories =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=  🐫
[default] synchronized from https://opam.ocaml.org
: mort@greyjay:papers#master *= $; opam upgrade
Everything as up-to-date as possible (run with --verbose to show unavailable upgrades).
The following would require downgrades or uninstalls, but you may upgrade them explicitly:
  - extlib.1.7.0
  - ocamlgraph.1.8.6
```

**Comment by jonludlam** [comment]
The failure looks to be in installing `userconf`, which is apparently already installed. You could try `ocamlfind remove userconf` and rerun?

**Comment by amirmc** [comment]
I tried removing `userconf` but now have different errors.
I'll try in a fresh switch and see if I can reproduce.

```
$ ocamlfind remove userconf
Removed /Users/amir/.opam/mirage/lib/userconf
$ opam install oasis
The following actions will be performed:
  ∗  install oasis 0.4.5

=-=- Gathering sources =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=  🐫
[oasis] Archive in cache

=-=- Processing actions -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=  🐫
[ERROR] The compilation of oasis failed at "ocaml setup.ml -install".
[WARNING] Directory /Users/amir/.opam/mirage/lib/oasis is not empty, not removing

#=== ERROR while installing oasis.0.4.5 =======================================#
# opam-version 1.2.2
# os           darwin
# command      ocaml setup.ml -install
# path         /Users/amir/.opam/mirage/build/oasis.0.4.5
# compiler     system (4.02.3)
# exit-code    1
# env-file     /Users/amir/.opam/mirage/build/oasis.0.4.5/oasis-47738-c61e7b.env
# stdout-file  /Users/amir/.opam/mirage/build/oasis.0.4.5/oasis-47738-c61e7b.out
# stderr-file  /Users/amir/.opam/mirage/build/oasis.0.4.5/oasis-47738-c61e7b.err
### stderr ###
# [...]
# Installed /Users/amir/.opam/mirage/lib/userconf/UserConf.cmi
# Installed /Users/amir/.opam/mirage/lib/userconf/UserConf.cmx
# Installed /Users/amir/.opam/mirage/lib/userconf/META
# W: Nothing to install for findlib library 'pluginloaderLib'
# W: Nothing to install for findlib library 'plugin3'
# W: Nothing to install for findlib library 'plugin2'
# W: Nothing to install for findlib library 'plugin1'
# ocamlfind: Package plugin-loader is already installed
#  - (file /Users/amir/.opam/mirage/lib/plugin-loader/META already exists)
# E: Failure("Command ''/Users/amir/.opam/mirage/bin/ocamlfind' install plugin-loader src/ext/plugin-loader/src/META _build/src/ext/plugin-loader/src/PluginLoader.cmx _build/src/ext/plugin-loader/src/PluginLoader.cmi _build/src/ext/plugin-loader/src/plugin-loader.cmxs _build/src/ext/plugin-loader/src/plugin-loader.a _build/src/ext/plugin-loader/src/plugin-loader.cmxa _build/src/ext/plugin-loader/src/plugin-loader.cma src/ext/plugin-loader/src/PluginLoader.ml' terminated with error code 2")
```

**Comment by amirmc** [comment]
Trying `opam install oasis` in a fresh switch doesn't have any problems. Not sure what's up with my `mirage` switch but I'll work in a different switch for now.  I'll leave this open in case there are other things I can try.

**Comment by verbosemode** [comment]
I ran into the same issue. However I can't reproduce it, since I've deleted $HOME/.opam/4.02.3/lib/userconf and then $HOME/.opam/4.02.3/lib/plugin-loader while troubleshooting. After these two directories were I gone I could install Oasis again without any problems.

**Comment by hcarty** [comment]
The problem seems to be with the removal of a previous installation of oasis.  I've run into the uninstall failure on multiple systems.

**Comment by amirmc** [comment]
Are there other issues we can cross-ref?  Would be useful for anyone else who hits this to be able to find some threads (until I posted this issue, I didn’t know anyone else had similar problems)

**Comment by gildor478** [comment]
@amirmc I think the problem is not directly related to OASIS but to a missing "uninstall" command somewhere in OPAM, I don't see what I can do about that.

**Comment by Chris00** [comment]
The oasis OPAM file was recently regenerated with more metadata — a bad thing retroactively — and this makes the uninstallation fail (the new script is used to remove the old package).  Just do 
```
ocamlfind remove plugin-loader
ocamlfind remove userconf
ocamlfind remove oasis
```
Will submit a fix.

**Comment by dsheets** [comment]
This was probably fixed by #5193.
